### ISSUE TITLE
`StronglyEntanglingLayers` applies the entangling gate twice when applied to two qubits

### ISSUE THREAD

**Issue by wongwsvincent** [PennyLaneAI/pennylane#1308]
#### Issue description

When ```qml.templates.layers.StronglyEntanglingLayers``` is called with two wires only, the 2-qubit gate (defined by the _imprimitive_ argument) is applied twice per layer. It would be nice to treat the special case of two wires only differently, like the ```qml.templates.layers.BasicEntanglerLayers``` feature for consistency.

Simple code for illustration:
```
import pennylane as qml
from pennylane.templates import StronglyEntanglingLayers
from pennylane.templates import BasicEntanglerLayers
from pennylane import numpy as np

method = 'StronglyEntanglingLayers'

dev = qml.device('default.qubit', wires=[0,1])

if(method=='StronglyEntanglingLayers'):
    shape = StronglyEntanglingLayers.shape(n_layers=1, n_wires=2)
elif(method=='BasicEntanglerLayers'):
    shape = BasicEntanglerLayers.shape(n_layers=1, n_wires=2)
weights = np.random.random(size=shape)

@qml.qnode(dev)
def circuit(params):
    if(method=='StronglyEntanglingLayers'):
        StronglyEntanglingLayers(params, wires=[0,1])
    elif(method=='BasicEntanglerLayers'):
        BasicEntanglerLayers(params, wires=[0,1])
    return qml.expval(qml.PauliZ(0))

drawer = qml.draw(circuit)
print(drawer(weights))
```
When ```StronglyEntanglingLayers``` is used:
```
 0: ──Rot(0.00845, 0.883, 0.23)──╭C──╭X──┤ ⟨Z⟩ 
 1: ──Rot(0.503, 0.814, 0.138)───╰X──╰C──┤     
```

When ```BasicEntanglerLayers``` is used:
```
0: ──RX(0.598)──╭C──┤ ⟨Z⟩ 
1: ──RX(0.749)──╰X──┤     
```

* Expected behavior: _2-qubit gate being applied once only per layer_

* Actual behavior: _2-qubit gate is applied twice instead_

* Reproduces how often: _Always_

* System information: (post the output of `import pennylane as qml; qml.about()`)
```
Name: PennyLane
Version: 0.15.1
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /Applications/anaconda3/envs/pennylane/lib/python3.6/site-packages
Requires: semantic-version, scipy, numpy, toml, networkx, autograd, appdirs
Required-by: 
Platform info:           Darwin-18.7.0-x86_64-i386-64bit
Python version:          3.6.10
Numpy version:           1.17.0
Scipy version:           1.5.0
Installed devices:
- default.gaussian (PennyLane-0.15.1)
- default.mixed (PennyLane-0.15.1)
- default.qubit (PennyLane-0.15.1)
- default.qubit.autograd (PennyLane-0.15.1)
- default.qubit.jax (PennyLane-0.15.1)
- default.qubit.tf (PennyLane-0.15.1)
- default.tensor (PennyLane-0.15.1)
- default.tensor.tf (PennyLane-0.15.1)
```

#### Source code and tracebacks
 ```StronglyEntanglingLayers``` source code (code of interest):
PennyLaneAI/pennylane/blob/master/pennylane/templates/layers/strongly_entangling.py#L114
 ```BasicEntanglerLayers ``` source code (for reference):
PennyLaneAI/pennylane/blob/master/pennylane/templates/layers/basic_entangler.py#L156

**Comment by mariaschuld** [comment]
Thanks for moving this from the discussion forum to an issue. I would be in favour of implementing this change, but don't know how to go about it: strictly speaking it would change the results of code that uses this template, but optimisation should be faster and I doubt the results in most application will differ much.

@co9olguy, shall we just implement the change and add a special "note" to the docs for one release cycle?

**Comment by wongwsvincent** [comment]
I have also just noticed that the custom _rangers_ option is never available for users to define. (```self.rangers``` is not defined if the argument ```rangers``` is not ```None```)
Please see [here](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/templates/layers/strongly_entangling.py#L89) for the relevant lines of code.

The implementation of user-defined rangers would require some value checks, for example, the ranger values cannot be divisible by the number of wires. If the sequence of rangers is shorter than the number of layers, I suppose we could repeat the sequence automatically.

Should I open a new issue, or should we fix the bug together here? 
I have a local working version, so I am happy to make a PR if needed.

**Comment by trbromley** [comment]
Thanks @wongwsvincent! Good point regarding the two-qubit version of `StronglyEntanglingLayers`. I agree that it might have been nicer to design it so that only one CNOT is applied as a special case for two qubits. However, now that we've been using the two-gate version for a while, I'm not sure there is much of an advantage to change now. The two-qubit setting is quick to simulate so the increased depth shouldn't matter. My only question would be if the circuit is less expressive/not universal :thinking:, in which case I'd then be supportive of changing.

> I have also just noticed that the custom ranges option is never available for users to define.

Good spot! Yes this looks like a bug. It would be great if you could make a new issue, and if you have a PR to fix it that would also be very welcome. Thanks!

**Comment by wongwsvincent** [comment]
Thanks @trbromley for the feedback! I myself (naively) don't see an advantage in expressivity for one over the other case. I brought the issue up purely for the cleanness of code and consistency with other functions, so I'm also fine if we prefer minimizing the number of insignificant changes and keeping it this way.

Regarding the custom range option, I have opened a new issue for it here: https://github.com/PennyLaneAI/pennylane/issues/1330

**Comment by mariaschuld** [comment]
Hey @wongwsvincent. We had an internal discussion and favour to keep the behaviour of `StronglyEmbeddingLayers` as it is (of course, except from the `ranges` fix - thanks for making a PR).

The reason is that the improvement is arguably small, but it is used a lot and we do not want to change the behaviour in everyone's code, even if the change would be for the better.

I will close the issue for now, but if you or anyone else wants to reopen it in future we are happy to pick up the discussion again!
